### ISSUE TITLE
False positive for `ref_in_deref` when extracting a function pointer field

### ISSUE THREAD

**Issue by kotauskas** [rust-lang/rust-clippy#6658]
Lint name: `ref_in_deref`

My procedural macro produced this code (manually reduced to MCVE):

```rust
// Input, retained in output
trait MyTrait {
    fn my_method(&self);
}
// end of input

#[derive(Copy, Clone, Debug, Hash)]
#[repr(C)]
struct MyTraitVtable {
    pub my_method: unsafe fn(*mut ()),
    pub drop: unsafe fn(*mut ()),
}

#[repr(transparent)]
struct BoxedMyTrait<'inner>(
    ::core::ptr::NonNull<MyTraitVtable>,
    ::core::marker::PhantomData<&'inner ()>,
);
impl MyTrait for BoxedMyTrait<'_> {
    fn my_method(&self) {
        unsafe {
            ((&mut *(self.0.as_ptr() as *mut MyTraitVtable)).my_method)(self as *const _ as *mut _)
        }
    }
}
```

Clippy produces the `ref_in_deref` lint on this line:
```rust
            ((&mut *(self.0.as_ptr() as *mut MyTraitVtable)).my_method)(self as *const _ as *mut _)
```
...and suggests to change it to this:
```rust
            (*(self.0.as_ptr() as *mut MyTraitVtable).my_method)(self as *const _ as *mut _)
```
The replacement produces a compilation error ("no field `my_method` on type `*mut hello::MyTraitVtable`").

### Meta

- `cargo clippy -V`: clippy 0.0.212 (`fe982319` 2020-11-19)
- `rustc -Vv`:
  ```
  rustc 1.50.0-nightly (fe982319a 2020-11-19)
  binary: rustc
  commit-hash: fe982319aa0aa5bbfc2795791a753832292bd2ba
  commit-date: 2020-11-19
  host: x86_64-pc-windows-msvc
  release: 1.50.0-nightly
  ```

**Comment by kotauskas** [comment]
Did a `rustup update`, no longer lints as of `clippy 0.1.51 (04caa632 2021-01-30)`.

**Comment by kotauskas** [comment]
Oops, evidently it still does.

**Comment by giraffate** [comment]
Thanks for your report!

I tried on [playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=26a1c196fea29d964a004b839bf532b0), but the suggested codes have no error. It seems to be fixed by https://github.com/rust-lang/rust-clippy/pull/6609, so I'm closing this. If you have any problems yet, feel free to reopen this.
